packages/dev/core/src/Meshes/csg2.ts

```diff
@@ -0,0 +1,513 @@
+                    vertProperties[i * numProp + offset + strideIndex] = component.data![i * component.stride + strideIndex];
+                }
+                offset += component.stride;
+            }
+        }
+
+        const manifoldMesh = new ManifoldMesh({ numProp: numProp, vertProperties, triVerts, runIndex, runOriginalID });
+        manifoldMesh.merge();
+
+        try {
+            return new CSG2(new Manifold(manifoldMesh), numProp, structure);
+        } catch (e) {
+            throw new Error("Error while creating the CSG: " + e.message);
+        }
+    }
+
+    /**
+     * Create a new Constructive Solid Geometry from a vertexData
+     * @param vertexData defines the vertexData to use to create the CSG
+     * @returns a new CSG2 class
+     */
+    public static FromVertexData(vertexData: VertexData): CSG2 {
+        const sourceVertices = vertexData.positions;
+        const sourceIndices = vertexData.indices;
+
+        if (!sourceVertices || !sourceIndices) {
+            throw new Error("The vertexData must at least have positions and indices");
+        }
+
+        // Create the MeshGL for I/O with Manifold library.
+        const triVerts = new Uint32Array(sourceIndices.length);
+
+        // Revert order
+        for (let i = 0; i < sourceIndices.length; i += 3) {
+            triVerts[i] = sourceIndices[i + 2];
+            triVerts[i + 1] = sourceIndices[i + 1];
+            triVerts[i + 2] = sourceIndices[i];
+        }
+
+        // Positions
+        let numProp = 3;
+        const structure: IManifoldVertexComponent[] = [{ stride: 3, kind: VertexBuffer.PositionKind, data: vertexData.positions! }];
+
+        // Normals
+        if (vertexData.normals) {
+            numProp += 3;
+            structure.push({ stride: 3, kind: VertexBuffer.NormalKind, data: vertexData.normals });
+        }
+
+        // UVs
+        for (const kind of [VertexBuffer.UVKind, VertexBuffer.UV2Kind, VertexBuffer.UV3Kind, VertexBuffer.UV4Kind, VertexBuffer.UV5Kind, VertexBuffer.UV6Kind]) {
+            const sourceUV = (vertexData as any)[kind === VertexBuffer.UVKind ? "uvs" : kind] as FloatArray;
+            if (sourceUV) {
+                numProp += 2;
+                structure.push({ stride: 2, kind: kind, data: sourceUV });
+            }
+        }
+
+        // Colors
+        const sourceColors = vertexData.colors;
+        if (sourceColors) {
+            numProp += 4;
+            structure.push({ stride: 4, kind: VertexBuffer.ColorKind, data: sourceColors });
+        }
+
+        return this._ProcessData(sourceVertices.length / 3, triVerts, structure, numProp);
+    }
+
+    /**
+     * Create a new Constructive Solid Geometry from a mesh
+     * @param mesh defines the mesh to use to create the CSG
+     * @param ignoreWorldMatrix defines if the world matrix should be ignored
+     * @returns a new CSG2 class
+     */
+    public static FromMesh(mesh: Mesh, ignoreWorldMatrix = false): any {

```

Comment:
Same thing here. However, we are about to set guidelines for this, so I guess we can wait with extracting these from the classes.